File: server/models/affiliate_profile.ts
import { DataTypes, Model, Optional, Sequelize } from 'sequelize';

interface AffiliateProfileAttributes {
  id: string;
  user_id: string;
  website?: string;
  social_media?: string;
  paypal_email?: string;
  bank_account?: string;
  commission_rate?: number;
}

// Making timestamps optional during creation
export interface AffiliateProfileCreationAttributes extends Optional<AffiliateProfileAttributes, 'id'> {}

export class AffiliateProfile extends Model<AffiliateProfileAttributes, AffiliateProfileCreationAttributes> implements AffiliateProfileAttributes {
  public id!: string;
  public user_id!: string;
  public website?: string;
  public social_media?: string;
  public paypal_email?: string;
  public bank_account?: string;
  public commission_rate?: number;

  // timestamps
  public readonly created_at!: Date;
  public readonly updated_at!: Date;

  public static associate(models: any): void {
    // Define association with User
    AffiliateProfile.belongsTo(models.User, {
      foreignKey: 'user_id',
      as: 'user'
    });
    
    // Define association with FlaggedActivity
    AffiliateProfile.hasMany(models.FlaggedActivity, {
      foreignKey: 'affiliate_id',
      as: 'flagged_activities',
      onDelete: 'SET NULL',
      onUpdate: 'CASCADE'
    });
  }
}

export function initAffiliateProfile(sequelize: Sequelize) {
  AffiliateProfile.init(
    {
      id: {
        type: DataTypes.STRING(36),
        primaryKey: true,
        defaultValue: DataTypes.UUIDV4,
      },
      user_id: {
        type: DataTypes.STRING(36),
        allowNull: false,
        references: { model: 'users', key: 'id' },
        onDelete: 'CASCADE',
      },
      website: DataTypes.STRING,
      social_media: DataTypes.STRING,
      paypal_email: DataTypes.STRING,
      bank_account: DataTypes.STRING,
      commission_rate: { type: DataTypes.DECIMAL(5, 2), defaultValue: 10.00 },
    },
    {
      sequelize,
      modelName: 'AffiliateProfile',
      tableName: 'affiliate_profiles',
      timestamps: true, // Let Sequelize manage timestamps
      underscored: true // Use snake_case for all field names
    }
  );

  AffiliateProfile.associate = function(models) {
    AffiliateProfile.belongsTo(models.User, {
      foreignKey: 'user_id',
      as: 'user',
    });
  };

  return AffiliateProfile;
}

File: server/models/affiliate_profile.py
import uuid
from sqlalchemy.sql import func
from . import db

class AffiliateProfile(db.Model):
    __tablename__ = 'affiliate_profiles'
    
    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = db.Column(db.String(36), db.ForeignKey('users.id', ondelete='CASCADE'), nullable=False)
    website = db.Column(db.String(255), nullable=True)
    social_media = db.Column(db.String(255), nullable=True)
    niche = db.Column(db.String(255), nullable=True)
    paypal_email = db.Column(db.String(255), nullable=True)
    bank_account = db.Column(db.String(255), nullable=True)
    commission_rate = db.Column(db.Numeric(5, 2), default=10.00, nullable=False)
    created_at = db.Column(db.DateTime, default=func.now(), nullable=False)
    updated_at = db.Column(db.DateTime, default=func.now(), onupdate=func.now(), nullable=False)
    
    # Relationships
    affiliate_links = db.relationship('AffiliateLink', backref='affiliate', lazy='dynamic', cascade='all, delete-orphan')
    
    def to_dict(self):
        """Convert affiliate profile object to dictionary"""
        return {
            'id': self.id,
            'user_id': self.user_id,
            'website': self.website,
            'social_media': self.social_media,
            'niche': self.niche,
            'paypal_email': self.paypal_email,
            'bank_account': self.bank_account,
            'commission_rate': float(self.commission_rate) if self.commission_rate is not None else None,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }
